### PR TITLE
Changed service definition + fixed testcases

### PR DESCRIPTION
**Pull request by jeroendk**
As mentioned in #27 I think it is better to use keys for the service definition, also most of the services already had keys so this is more consistent. 

In addition i fixed the failing tests.

### CODE REVIEW

**Comment by jeroendk** [comment]
@Nightbr Phpsepc still fails because the function subscriberHash is now static in DrewM\MailChimp. Any idea on how to fix this?

File: src/Resources/config/services.yml [hunk 1]
```diff
@@ -52,14 +52,14 @@ services:
             - { name: kernel.event_listener, event: welp.mailchimp.delete, method: onDelete }
 
     # Commands
-    Welp\MailchimpBundle\Command\SynchronizeMergeFieldsCommand:
+    welp_mailchimp.command.synchronize_merge_fields:
```
Comment (Nightbr) [inline src/Resources/config/services.yml:55]: If you wanbt to use id, you have to set class in the definition. 
For example:
```
welp_mailchimp.command.synchronize_subscribers:
  class: Welp\MailchimpBundle\Command\SynchronizeSubscribersCommand
  tags:
     - { name: 'console.command', command: 'welp:mailchimp:synchronize-subscribers' }   
```

**Review by Nightbr — CHANGES_REQUESTED** [review]
If you want to use ids for services, you have to set class manually (no autowire here).

For the problem with [drewm/mailchimp-api](https://github.com/drewm/mailchimp-api). You need to replace in `ListRepository.php`:
`$this->mailchimp->subscriberHash` by  `MailChimp::subscriberHash` because subscriberHash became a static method.

After that, you need to fix ListRepositoryTest.php (no time for this, sorry, but if you can manage it, we will be able to merge this and release a version :+1: ).

```
Welp/MailchimpBundle/Subscriber/ListRepository                          
  48  - it subscribe a subscriber
      unexpected method call on Double\DrewM\MailChimp\MailChimp\P34:
        - put(
              "lists/ba039c6198/members/b1a29fd58778c40c7f15f06a334dc691",
              ["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "subscribed"]
          )
      expected calls were:
        - subscriberHash(
              exact("charles@terrasse.fr")
          )
        - success(
      
          )
        - get(
              exact("lists/ba039c6198")
          )
        - get(
              exact("lists/notfound")
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "subscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "unsubscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "pending"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "cleaned"])
          )

Welp/MailchimpBundle/Subscriber/ListRepository                        
  59  - it unsubscribe a subscriber
      unexpected method call on Double\DrewM\MailChimp\MailChimp\P36:
        - put(
              "lists/ba039c6198/members/b1a29fd58778c40c7f15f06a334dc691",
              ["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "unsubscribed"]
          )
      expected calls were:
        - subscriberHash(
              exact("charles@terrasse.fr")
          )
        - success(
      
          )
        - get(
              exact("lists/ba039c6198")
          )
        - get(
              exact("lists/notfound")
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "subscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "unsubscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "pending"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "cleaned"])
          )

Welp/MailchimpBundle/Subscriber/ListRepository                        
  64  - it pending a subscriber
      unexpected method call on Double\DrewM\MailChimp\MailChimp\P38:
        - put(
              "lists/ba039c6198/members/b1a29fd58778c40c7f15f06a334dc691",
              ["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "pending"]
          )
      expected calls were:
        - subscriberHash(
              exact("charles@terrasse.fr")
          )
        - success(
      
          )
        - get(
              exact("lists/ba039c6198")
          )
        - get(
              exact("lists/notfound")
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "subscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "unsubscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "pending"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "cleaned"])
          )

Welp/MailchimpBundle/Subscriber/ListRepository                        
  69  - it clean a subscriber
      unexpected method call on Double\DrewM\MailChimp\MailChimp\P40:
        - put(
              "lists/ba039c6198/members/b1a29fd58778c40c7f15f06a334dc691",
              ["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "cleaned"]
          )
      expected calls were:
        - subscriberHash(
              exact("charles@terrasse.fr")
          )
        - success(
      
          )
        - get(
              exact("lists/ba039c6198")
          )
        - get(
              exact("lists/notfound")
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "subscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "unsubscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "pending"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "cleaned"])
          )

Welp/MailchimpBundle/Subscriber/ListRepository                        
  74  - it delete a subscriber
      unexpected method call on Double\DrewM\MailChimp\MailChimp\P42:
        - delete(
              "lists/ba039c6198/members/b1a29fd58778c40c7f15f06a334dc691"
          )
      expected calls were:
        - subscriberHash(
              exact("charles@terrasse.fr")
          )
        - success(
      
          )
        - get(
              exact("lists/ba039c6198")
          )
        - get(
              exact("lists/notfound")
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "subscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "unsubscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "pending"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "cleaned"])
          )
        - delete(
              exact("lists/ba039c6198/members/md5ofthesubscribermail")
          )

Welp/MailchimpBundle/Subscriber/ListRepository                        
  81  - it update a subscriber
      unexpected method call on Double\DrewM\MailChimp\MailChimp\P44:
        - patch(
              "lists/ba039c6198/members/b1a29fd58778c40c7f15f06a334dc691",
              ["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html"]
          )
      expected calls were:
        - subscriberHash(
              exact("charles@terrasse.fr")
          )
        - success(
      
          )
        - get(
              exact("lists/ba039c6198")
          )
        - get(
              exact("lists/notfound")
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "subscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "unsubscribed"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "pending"])
          )
        - put(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html", "status" => "cleaned"])
          )
        - patch(
              exact("lists/ba039c6198/members/md5ofthesubscribermail"),
              exact(["email_address" => "charles@terrasse.fr", "merge_fields" => ["FNAME" => "Charles", "LNAME" => "Terrasse"], "language" => "fr", "email_type" => "html"])
          )

```

**Comment by Nightbr** [comment]
Add to that, I think we should lock the version of drewm/mailchimp-api to 2.5.* to avoid this sort of thing in the future.

**composer.json**
`"drewm/mailchimp-api": "2.5.*"`

File: src/Resources/config/services.yml (hunk 1 above)
Comment (jeroendk) [reply src/Resources/config/services.yml:55]: Yes obviously! my bad...

**Comment by jeroendk** [comment]
So i changed "md5ofthesubscribermail" for the md5 hash of the test email because you can not mock a static method.  In there Github they suggest to create your own wrapper/proxy but that might be a little to much for this one thing? 
https://github.com/phpspec/prophecy/pull/20#issuecomment-18133965

**Review by Nightbr — APPROVED** [review]
👍 Thanks for your contribution 🎉 
I will merge this and ship a release soon :)

**Comment by Nightbr** [comment]
https://github.com/welpdev/mailchimp-bundle/releases/tag/1.3.0

**Pull request merged**